expression-language.md

```diff
@@ -0,0 +1,369 @@
+# CloudEvents SQL Expression Language
+
+## Abstract
+
+The goal of this specification is to define an expression language SQL-like which can be used to express predicates on
+CloudEvents instances.
+
+## Status of this document
+
+This document is a working draft.
+
+## Table of Contents
+
+1. [Introduction](#1-introduction)
+
+- 1.1. [Conformance](#11-conformance)
+- 1.2. [Relation to Subscriptions API](#12-relation-to-the-subscriptions-api)
+
+2. [Language syntax](#2-language-syntax)
+
+- 2.1. [Expression](#21-expression)
+- 2.2. [Value identifiers and literals](#22-value-identifiers-and-literals)
+- 2.3. [Operators](#23-operators)
+- 2.4. [Functions invocation](#24-functions-invocation)
+
+3. [Language semantics](#3-language-semantics)
+
+- 3.1. [Type System](#31-type-system)
+- 3.2. [CloudEvent context identifiers and types](#32-cloudevent-context-identifiers-and-types)
+- 3.3. [Errors](#33-errors)
+- 3.4. [Operators](#34-operators)
+- 3.5. [Functions](#35-functions)
+- 3.6. [Evaluation of the expression](#36-evaluation-of-the-expression)
+
+4. [Examples](#4-examples)
+5. [References](#5-references)
+
+## 1. Introduction
+
+CloudEvents SQL expressions (also known as CESQL) allow matching of CloudEvent attributes against complex expressions
+that lean on the syntax of Structured Query Language (SQL) `WHERE` clauses. Using SQL-derived expressions for message
+filtering is in widespread implementation use because the Java Message Service (JMS) message selector syntax also leans
+on SQL. Note that neither the SQL standard (ISO 9075) nor the JMS standard nor any other SQL dialect are used as a
+normative foundation or constrain the expression syntax defined in this specification, but the syntax is informed by
+them.
+
+CESQL is a _[Total pure functional programming language][total-programming-language-wiki]_ in order to guarantee the
+termination of the evaluation of the expression. It features a type system correlated to the [CloudEvents type
+system][ce-type-system] and it features boolean and arithmetics operations, as well as built-in functions for string
+manipulation.
+
+The language is not constrained to a particular execution environment, which means it might run in a source, in a
+producer, in an intermediary, and it can be implemented using any tech stack.
+
+The CloudEvents Expression Language assumes the input always includes, but it's not limited to, a single valid and type
+checked CloudEvent instance. An expression cannot mutate the value of the input CloudEvent instance, nor any of the
+other input values. The evaluation of an expression observes the concept of [referential
+transparency][referential-transparency-wiki].
+
+The CloudEvents Expression Language doesn't support the handling of the data field of the CloudEvent instances, due to
+its polymorphic nature and complexity. We strongly encourage users that needs this functionality to use other more
```

Comment:
From a user perspective it could be useful for some cases like json.

Reply:
I see it, but as already discussed in previous Serverless WG meetings, it's just too hard to filter on the body. That's also something that middlewares might not want to do, because it's generally heavy.

Also note that for body filtering, depending on the "body nature", there might be tools better suited for that purpose: e.g. to filter a json payload a tool like json path/jsoniq definitely fits better than this language